File: app/src/main/java/com/veganbeauty/app/features/shop/search/ShopSearchDetailFragment.java
package com.veganbeauty.app.features.shop.search;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.veganbeauty.app.R;
import com.veganbeauty.app.core.base.RootieFragment;
import com.veganbeauty.app.data.local.LocalJsonReader;
import com.veganbeauty.app.data.local.RootieDatabase;
import com.veganbeauty.app.data.local.entities.ProductEntity;
import com.veganbeauty.app.data.repository.ProductRepository;
import com.veganbeauty.app.databinding.ShopSearchDetailBinding;
import com.veganbeauty.app.features.shop.product.detail.ProductDetailLauncher;
import com.veganbeauty.app.features.shop.product.list.ShopListFragment;

public class ShopSearchDetailFragment extends RootieFragment {

    private ShopSearchDetailBinding binding;

    private ShopSearchViewModel searchViewModel;
    private SearchSuggestionAdapter keywordSuggestionAdapter;
    private SearchContentSuggestionAdapter contentSuggestionAdapter;
    private HotDealsAdapter previewProductsAdapter;

    private static final String ARG_KEYWORD = "arg_keyword";

    public static ShopSearchDetailFragment newInstance(String keyword) {
        ShopSearchDetailFragment fragment = new ShopSearchDetailFragment();
        Bundle args = new Bundle();
        args.putString(ARG_KEYWORD, keyword);
        fragment.setArguments(args);
        return fragment;
    }

    public static ShopSearchDetailFragment newInstance() {
        return newInstance("");
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = ShopSearchDetailBinding.inflate(inflater, container, false);
        setupViewModel();
        return binding.getRoot();
    }

    private void setupViewModel() {
        RootieDatabase db = RootieDatabase.getDatabase(requireContext());
        ProductRepository repository = new ProductRepository(db.productDao(), new LocalJsonReader(requireContext()));
        searchViewModel = new ViewModelProvider(this, new ViewModelProvider.Factory() {
            @NonNull
            @Override
            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
                return (T) new ShopSearchViewModel(repository, db.communityDao());
            }
        }).get(ShopSearchViewModel.class);
    }

    @Override
    public void setupUI(@NonNull View view) {
        binding.btnBack.setOnClickListener(v -> getParentFragmentManager().popBackStack());
        binding.btnClear.setOnClickListener(v -> binding.etSearch.getText().clear());

        keywordSuggestionAdapter = new SearchSuggestionAdapter(R.drawable.ic_search, name -> {
            binding.etSearch.setText(name);
            binding.etSearch.setSelection(name.length());
            performSearch();
        });
        binding.rvKeywordSuggestions.setAdapter(keywordSuggestionAdapter);
        binding.rvKeywordSuggestions.setLayoutManager(new LinearLayoutManager(requireContext()));

        contentSuggestionAdapter = new SearchContentSuggestionAdapter(this::handleContentSuggestionClick);
        binding.rvCategorySuggestions.setAdapter(contentSuggestionAdapter);
        binding.rvCategorySuggestions.setLayoutManager(new LinearLayoutManager(requireContext()));

        previewProductsAdapter = new HotDealsAdapter(
                this::navigateToDetail,
                product -> {
                    com.veganbeauty.app.features.shop.product.ChooseQuantityBottomSheet bottomSheet = new com.veganbeauty.app.features.shop.product.ChooseQuantityBottomSheet(product, new com.veganbeauty.app.features.shop.product.ChooseQuantityBottomSheet.OnQuantitySelectedListener() {
                        @Override
                        public void onAddToCartClick(com.veganbeauty.app.data.local.entities.ProductEntity p, int quantity) {
                            com.veganbeauty.app.features.shop.product.CartHelper.addToCart(requireContext(), androidx.lifecycle.LifecycleOwnerKt.getLifecycleScope(ShopSearchDetailFragment.this), p, quantity);
                        }
                        @Override
                        public void onBuyNowClick(com.veganbeauty.app.data.local.entities.ProductEntity p, int quantity) {}
                    });
                    bottomSheet.show(getParentFragmentManager(), "ChooseQuantity");
                }
        );
        binding.rvSearchResults.setAdapter(previewProductsAdapter);
        binding.rvSearchResults.setLayoutManager(new LinearLayoutManager(requireContext()));

        String initialKeyword = "";
        if (getArguments() != null) {
            initialKeyword = getArguments().getString(ARG_KEYWORD, "");
        }
        if (!initialKeyword.isEmpty()) {
            binding.etSearch.setText(initialKeyword);
            binding.etSearch.setSelection(initialKeyword.length());
            searchViewModel.updateSuggestions(initialKeyword);
        }

        binding.etSearch.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                performSearch();
                return true;
            }
            return false;
        });

        binding.etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                String keyword = s != null ? s.toString().trim() : "";
                binding.btnClear.setVisibility(keyword.isEmpty() ? View.GONE : View.VISIBLE);
                searchViewModel.updateSuggestions(keyword);
            }
        });
    }

    @Override
    protected void observeViewModel() {
        searchViewModel.suggestions.observe(getViewLifecycleOwner(), suggestions -> {
            keywordSuggestionAdapter.submitList(suggestions.getProductNames());
            contentSuggestionAdapter.submitList(suggestions.getContentItems());
            previewProductsAdapter.submitList(suggestions.getPreviewProducts());

            binding.rvKeywordSuggestions.setVisibility(
                    suggestions.getProductNames().isEmpty() ? View.GONE : View.VISIBLE);
            binding.dividerContent.setVisibility(
                    (suggestions.getProductNames().isEmpty() || suggestions.getContentItems().isEmpty()) ? View.GONE : View.VISIBLE);
            binding.rvCategorySuggestions.setVisibility(
                    suggestions.getContentItems().isEmpty() ? View.GONE : View.VISIBLE);
            binding.dividerProducts.setVisibility(
                    (suggestions.getContentItems().isEmpty() && suggestions.getProductNames().isEmpty()) ? View.GONE
                            : (suggestions.getPreviewProducts().isEmpty() ? View.GONE : View.VISIBLE));
            binding.rvSearchResults.setVisibility(
                    suggestions.getPreviewProducts().isEmpty() ? View.GONE : View.VISIBLE);
        });
    }

    private void handleContentSuggestionClick(ContentSuggestion item) {
        if (item.getType() == ContentSuggestionType.CATEGORY) {
            String parentCategory = item.getParentCategory();
            String category = item.getCategoryName();
            if (parentCategory == null && category == null) return;
            if (parentCategory != null) {
                navigateToCategoryList(parentCategory, category);
            } else {
                navigateToCategoryList(category, null);
            }
        } else if (item.getType() == ContentSuggestionType.VIDEO) {
            if (item.getVideoUrl() != null) {
                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(item.getVideoUrl())));
            }
        } else if (item.getType() == ContentSuggestionType.BLOG || item.getType() == ContentSuggestionType.POST) {
            performSearch();
        }
    }

    private void performSearch() {
        Editable text = binding.etSearch.getText();
        String keyword = text != null ? text.toString().trim() : "";
        if (keyword.isEmpty()) return;
        ShopSearchHistoryHelper.add(requireContext(), keyword);
        getParentFragmentManager().beginTransaction()
                .replace(R.id.main_container, ShopSearchResultsFragment.newInstance(keyword))
                .addToBackStack(null)
                .commit();
    }

    private void navigateToCategoryList(String category, String subcategory) {
        ShopListFragment listFragment = new ShopListFragment();
        Bundle args = new Bundle();
        args.putString("CATEGORY_NAME", category);
        if (subcategory != null) {
            args.putString("SUBCATEGORY_NAME", subcategory);
        }
        listFragment.setArguments(args);
        getParentFragmentManager().beginTransaction()
                .replace(R.id.main_container, listFragment)
                .addToBackStack(null)
                .commit();
    }

    private void navigateToDetail(ProductEntity product) {
        ProductDetailLauncher.open(this, product);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/veganbeauty/app/features/shop/search/ShopSearchDetailFragment.kt
package com.veganbeauty.app.features.shop.search

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.veganbeauty.app.R
import com.veganbeauty.app.core.base.RootieFragment
import com.veganbeauty.app.data.local.LocalJsonReader
import com.veganbeauty.app.data.local.RootieDatabase
import com.veganbeauty.app.data.local.entities.ProductEntity
import com.veganbeauty.app.data.repository.ProductRepository
import com.veganbeauty.app.databinding.ShopSearchDetailBinding
import com.veganbeauty.app.features.shop.product.detail.ShopDetailFragment
import com.veganbeauty.app.features.shop.product.list.ShopListFragment

class ShopSearchDetailFragment : RootieFragment() {

    private var _binding: ShopSearchDetailBinding? = null
    private val binding get() = _binding!!

    private lateinit var searchViewModel: ShopSearchViewModel
    private lateinit var keywordSuggestionAdapter: SearchSuggestionAdapter
    private lateinit var contentSuggestionAdapter: SearchContentSuggestionAdapter
    private lateinit var previewProductsAdapter: HotDealsAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = ShopSearchDetailBinding.inflate(inflater, container, false)
        setupViewModel()
        return binding.root
    }

    private fun setupViewModel() {
        val db = RootieDatabase.getDatabase(requireContext())
        val repository = ProductRepository(db.productDao(), LocalJsonReader(requireContext()))
        searchViewModel = ViewModelProvider(this, object : ViewModelProvider.Factory {
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                @Suppress("UNCHECKED_CAST")
                return ShopSearchViewModel(repository, db.communityDao()) as T
            }
        })[ShopSearchViewModel::class.java]
    }

    override fun setupUI(view: View) {
        binding.btnBack.setOnClickListener { parentFragmentManager.popBackStack() }
        binding.btnClear.setOnClickListener { binding.etSearch.text.clear() }

        keywordSuggestionAdapter = SearchSuggestionAdapter(R.drawable.ic_search) { name ->
            binding.etSearch.setText(name)
            binding.etSearch.setSelection(name.length)
            performSearch()
        }
        binding.rvKeywordSuggestions.adapter = keywordSuggestionAdapter
        binding.rvKeywordSuggestions.layoutManager = LinearLayoutManager(requireContext())

        contentSuggestionAdapter = SearchContentSuggestionAdapter { item ->
            handleContentSuggestionClick(item)
        }
        binding.rvCategorySuggestions.adapter = contentSuggestionAdapter
        binding.rvCategorySuggestions.layoutManager = LinearLayoutManager(requireContext())

        previewProductsAdapter = HotDealsAdapter(onItemClick = { navigateToDetail(it) })
        binding.rvSearchResults.adapter = previewProductsAdapter
        binding.rvSearchResults.layoutManager = LinearLayoutManager(requireContext())

        val initialKeyword = arguments?.getString(ARG_KEYWORD).orEmpty()
        if (initialKeyword.isNotEmpty()) {
            binding.etSearch.setText(initialKeyword)
            binding.etSearch.setSelection(initialKeyword.length)
            searchViewModel.updateSuggestions(initialKeyword)
        }

        binding.etSearch.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                performSearch()
                true
            } else false
        }

        binding.etSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                val keyword = s?.toString()?.trim().orEmpty()
                binding.btnClear.visibility = if (keyword.isEmpty()) View.GONE else View.VISIBLE
                searchViewModel.updateSuggestions(keyword)
            }
        })
    }

    override fun observeViewModel() {
        searchViewModel.suggestions.observe(viewLifecycleOwner) { suggestions ->
            keywordSuggestionAdapter.submitList(suggestions.productNames)
            contentSuggestionAdapter.submitList(suggestions.contentItems)
            previewProductsAdapter.submitList(suggestions.previewProducts)

            binding.rvKeywordSuggestions.visibility =
                if (suggestions.productNames.isEmpty()) View.GONE else View.VISIBLE
            binding.dividerContent.visibility =
                if (suggestions.productNames.isEmpty() || suggestions.contentItems.isEmpty()) View.GONE else View.VISIBLE
            binding.rvCategorySuggestions.visibility =
                if (suggestions.contentItems.isEmpty()) View.GONE else View.VISIBLE
            binding.dividerProducts.visibility =
                if (suggestions.contentItems.isEmpty() && suggestions.productNames.isEmpty()) View.GONE
                else if (suggestions.previewProducts.isEmpty()) View.GONE else View.VISIBLE
            binding.rvSearchResults.visibility =
                if (suggestions.previewProducts.isEmpty()) View.GONE else View.VISIBLE
        }
    }

    private fun handleContentSuggestionClick(item: ContentSuggestion) {
        when (item.type) {
            ContentSuggestionType.CATEGORY -> {
                val category = item.categoryName ?: return
                navigateToCategoryList(category, item.subcategoryName)
            }
            ContentSuggestionType.VIDEO -> {
                item.videoUrl?.let { startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(it))) }
            }
            ContentSuggestionType.BLOG, ContentSuggestionType.POST -> performSearch()
        }
    }

    private fun performSearch() {
        val keyword = binding.etSearch.text?.toString()?.trim().orEmpty()
        if (keyword.isEmpty()) return
        ShopSearchHistoryHelper.add(requireContext(), keyword)
        parentFragmentManager.beginTransaction()
            .replace(R.id.main_container, ShopSearchResultsFragment.newInstance(keyword))
            .addToBackStack(null)
            .commit()
    }

    private fun navigateToCategoryList(category: String, subcategory: String? = null) {
        val listFragment = ShopListFragment().apply {
            arguments = Bundle().apply {
                putString("CATEGORY_NAME", category)
                subcategory?.let { putString("SUBCATEGORY_NAME", it) }
            }
        }
        parentFragmentManager.beginTransaction()
            .replace(R.id.main_container, listFragment)
            .addToBackStack(null)
            .commit()
    }

    private fun navigateToDetail(product: ProductEntity) {
        parentFragmentManager.beginTransaction()
            .replace(R.id.main_container, ShopDetailFragment().apply { setProduct(product) })
            .addToBackStack(null)
            .commit()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        private const val ARG_KEYWORD = "arg_keyword"

        fun newInstance(keyword: String = ""): ShopSearchDetailFragment {
            return ShopSearchDetailFragment().apply {
                arguments = Bundle().apply { putString(ARG_KEYWORD, keyword) }
            }
        }
    }
}
